src/video/texture_atlas.c

```diff
@@ -0,0 +1,65 @@
+#include <string.h>
+
+#include "utils/hashmap.h"
+#include "video/texture_atlas.h"
+
+// WIP
+
+typedef struct {
+    uint64_t id;
+    uint16_t x;
+    uint16_t y;
+    uint16_t w;
+    uint16_t h;
+} atlas_item;
+
+typedef struct texture_atlas_t {
+    hashmap items;
+    GLuint texture_id;
+    uint16_t w;
+    uint16_t h;
+} texture_atlas;
+
+static GLuint create_atlas_texture(uint16_t width, uint16_t height) {
+    GLuint texture_id = 0;
+    glGenTextures(1, &texture_id);
+    glBindTexture(GL_TEXTURE_2D, texture_id);
+    glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RG_INTEGER, width, height, 0, GL_RG_INTEGER, GL_UNSIGNED_BYTE, NULL);
```

Comment:
Prefer `TexStorage` functions over `TexImage`. Need to check extension is supported. Though there are no modern platforms with supported OpenGL where it isn't. Maybe WebGL

Either generate and set mipmaps or set max miplevel to 0.

Reply:
Not sure if `GL_RG_INTEGER` is a guaranteed supported type. Might have to check that first.

Reply:
Yeah, none of this is likely to work. I just copied my ancient GL2.1 tex atlas for a starting point and threw in some ideas. I may make a separate stencil for the sprites to handle the alpha too, we shall see. Also, it might be better to copy the sprite bytecodes to a buffer and then just render the sprites from that. We shall see.

Reply:
Righto, the bytecode was too complicated for a shader to do easily, maybe just a texture atlas then. TexStorage* stuff seems to be GL 4.2... I suppose I could use those regarless as you said, but I just don't want to deal with it now :D Maybe an optional future path for it, if it is supported.

For texture data, I need to dump the shaders paletted textures so 1 byte of image data, and then some sort of 1 bit alpha value (on/off). Might be better to hold two textures, one with GL_RED for images and then something else for the alpha (I wonder if GL has something for black/white stuff that could be used for that alpha).